File: src/taeeun/week2/2-2.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    map<int, int> a;
    map<int, int> b;
    a[topping[0]]++;
    for (int i=1; i<topping.size(); i++) {
       b[topping[i]]++;
    }
    for (int i=1; i<topping.size()-1; i++) {
        if (a.size()==b.size()) answer++;
        a[topping[i]]++;
        b[topping[i]]--;
        if (b[topping[i]]==0) b.erase(topping[i]);
    }
    return answer;
}

File: src/geonhyeong/week2/문제집 2/2-2.py
#
#  2-2.swift
#  Question Link:
#  Primary idea: dict
#
#  문제 시간 : 약 30분
#  Time Complexity : O()
#  Space Complexity : O()
#  Runtime:  ms
#  Memory Usage:  MB
#

from collections import defaultdict

def solution(topping):
    answer = 0
    A_set = set()
    B_dict = defaultdict(int)

    # 초기화
    for t in topping:
        B_dict[t] += 1
    
    # 종류 갯수
    A_count = 0
    B_count = len(B_dict)

    for t in topping:
        if not t in A_set:
            A_set.add(t)
            A_count += 1
        
        B_dict[t] -= 1
        if B_dict[t] == 0:
            B_count -= 1

        if A_count == B_count:
            answer += 1

    return answer

print(solution([1, 2, 1, 3, 1, 4, 1, 2])) # 2
# print(solution([1, 2, 3, 1, 4])) # 0
